File: skills/tokensmind-agent-network-runtime/scripts/action_executor.py
import uuid

from action_support import ActionState, AuthorizationRequired
from action_validation import text
from python_runtime.errors import AgentNetworkHttpError


def validate_request(value, operations):
    operation = text(value.get("operation")) if isinstance(value, dict) else ""
    data = value.get("input") or {} if isinstance(value, dict) else {}
    if operation not in operations or not isinstance(data, dict):
        raise ActionState(
            "input_required",
            "Provide a supported operation and object input.",
            fields=["operation", "input"],
        )
    return operation, data


def _claim(workflow, operation, data):
    current = workflow.read()
    if not current:
        candidate = {
            "version": 1,
            "id": str(uuid.uuid4()),
            "operation": operation,
            "input": data,
        }
        workflow.write(candidate)
        current = workflow.read()
    if not current:
        raise RuntimeError("Workflow claim returned no state")
    if current.get("operation") != operation or current.get("input") != data:
        raise ActionState(
            "failed",
            "Another %s action is pending." % current.get("operation", "Agent Network"),
            code="AGENT_NETWORK_ACTION_PENDING",
            retryable=False,
        )
    return current


def _abandon(workflow):
    abandoned = workflow.remove()
    return {
        "status": "completed",
        "operation": "abandon_action",
        "data": {"abandoned": abandoned},
    }


def _failure_result(error):
    is_http = isinstance(error, AgentNetworkHttpError)
    retryable = is_http and (error.status == 429 or error.status >= 500)
    result = {
        "status": "failed",
        "code": getattr(error, "code", None) or "AGENT_NETWORK_ACTION_ERROR",
        "message": str(error),
        "retryable": retryable,
    }
    if not is_http or not isinstance(error.status, int):
        return result
    result["httpStatus"] = error.status
    if error.correction:
        result["correction"] = error.correction
    if error.retry_after:
        result["retryAfter"] = error.retry_after
    return result


class ActionExecutor:
    def __init__(self, api, workflow, *, handlers, operations):
        self._api = api
        self._workflow = workflow
        self._handlers = handlers
        self._operations = operations

    def execute(self, value):
        try:
            operation, data = validate_request(value, self._operations)
        except ActionState as error:
            return {"status": error.status, **error.data}
        if operation == "abandon_action":
            return _abandon(self._workflow)
        try:
            current = _claim(self._workflow, operation, data)
        except ActionState as error:
            return {"status": error.status, **error.data}
        except Exception as error:
            return _failure_result(error)
        return self._run(operation, data, current["id"])

    def _run(self, operation, data, key):
        try:
            result = self._handlers[operation](self._api, data, key)
            self._workflow.remove()
            return {"status": "completed", "operation": operation, "data": result}
        except ActionState as error:
            if error.status != "authorization_required":
                self._workflow.remove()
            return {"status": error.status, **error.data}
        except AuthorizationRequired as error:
            return {
                "status": "authorization_required",
                "message": str(error),
                "verificationUrl": error.url,
            }
        except Exception as error:
            result = _failure_result(error)
            if not result["retryable"]:
                self._workflow.remove()
            return result

File: skills/tokensmind-agent-network/scripts/lib/action-executor.js
import { randomUUID } from 'node:crypto';
import { ActionState, serializeFailure } from './action-errors.js';
import { createActionContext } from './action-context.js';
import { validateActionRequest } from './action-validation.js';
import { ensureAgent, getMyAgent, searchAgents, updateAgent } from './agent-actions.js';
import { contactAgent } from './contact-action.js';
import { appeal, blockAgent, report, unblockAgent } from './governance-actions.js';
import {
  getConversation,
  listInbox,
  markRead,
  reply,
  withdrawMessage,
} from './messaging-actions.js';
import {
  deleteMemory,
  getMemorySettings,
  listMemories,
  proposeMemory,
} from './memory-actions.js';

const ACTIONS = Object.freeze({
  appeal,
  block_agent: blockAgent,
  contact_agent: contactAgent,
  ensure_agent: ensureAgent,
  update_agent: updateAgent,
  get_conversation: getConversation,
  get_my_agent: getMyAgent,
  list_inbox: listInbox,
  mark_read: markRead,
  reply,
  report,
  search_agents: searchAgents,
  unblock_agent: unblockAgent,
  withdraw_message: withdrawMessage,
  delete_memory: deleteMemory,
  get_memory_settings: getMemorySettings,
  list_memories: listMemories,
  propose_memory: proposeMemory,
});

function canonicalize(value) {
  if (Array.isArray(value)) return value.map(canonicalize);
  if (!value || typeof value !== 'object') return value;
  return Object.fromEntries(
    Object.keys(value).sort().map((key) => [key, canonicalize(value[key])]),
  );
}

function requestsMatch(workflow, request) {
  return JSON.stringify(canonicalize({
    operation: workflow.operation,
    input: workflow.input,
  })) === JSON.stringify(canonicalize(request));
}

async function loadWorkflow({ request, store, randomId }) {
  let current = await store.read('workflow');
  if (!current) {
    const candidate = { version: 1, id: randomId(), ...request };
    await store.write('workflow', candidate);
    current = await store.read('workflow');
  }
  if (!current) throw new Error('Workflow claim returned no state');
  if (current && !requestsMatch(current, request)) {
    const error = new Error(`Another ${current.operation} action is pending.`);
    error.code = 'AGENT_NETWORK_ACTION_PENDING';
    throw error;
  }
  return current;
}

async function handleFailure(error, store) {
  if (error instanceof ActionState) {
    if (error.status !== 'authorization_required') await store.remove('workflow');
    return { status: error.status, ...error.data };
  }
  if (error?.status === 'authorization_required') {
    return {
      status: 'authorization_required',
      message: error.message,
      verificationUrl: error.verificationUrl,
    };
  }
  const serialized = serializeFailure(error);
  if (!serialized.retryable) await store.remove('workflow');
  return { status: 'failed', ...serialized };
}

function serializeWithoutCleanup(error) {
  if (error instanceof ActionState) return { status: error.status, ...error.data };
  return { status: 'failed', ...serializeFailure(error) };
}

async function abandonWorkflow(store) {
  const abandoned = await store.remove('workflow');
  return {
    status: 'completed',
    operation: 'abandon_action',
    data: { abandoned },
  };
}

export function createActionExecutor({ api, workflowStore, randomId = randomUUID }) {
  return {
    async execute(value) {
      let request;
      try {
        request = validateActionRequest(value);
      } catch (error) {
        return serializeWithoutCleanup(error);
      }
      if (request.operation === 'abandon_action') {
        try {
          return await abandonWorkflow(workflowStore);
        } catch (error) {
          return serializeWithoutCleanup(error);
        }
      }
      let workflow;
      try {
        workflow = await loadWorkflow({ request, store: workflowStore, randomId });
      } catch (error) {
        return serializeWithoutCleanup(error);
      }
      try {
        const context = createActionContext({ api, workflow });
        const data = await ACTIONS[request.operation](context);
        await workflowStore.remove('workflow');
        return { status: 'completed', operation: request.operation, data };
      } catch (error) {
        return handleFailure(error, workflowStore);
      }
    },
  };
}
